src/robot/variables/replacer.py

```diff
@@ -137,46 +137,65 @@ def _get_variable_item(self, match, value):
+    def _get_sequence_variable_item(self, name, variable, index):
         index = self.replace_string(index)
         try:
-            index = self._parse_list_variable_index(index, name[0] == '$')
+            index = self._parse_sequence_variable_index(index, name[0] == '$')
         except ValueError:
-            raise VariableError("List '%s' used with invalid index '%s'."
-                                % (name, index))
+            raise VariableError("%s '%s' used with invalid index '%s'. "
+                                "To use '[%s]' as a literal value, it needs "
+                                "to be escaped like '\\[%s]'."
+                                % (type_name(variable, capitalize=True), name,
+                                   index, index, index))
         try:
             return variable[index]
         except IndexError:
-            raise VariableError("List '%s' has no item in index %d."
-                                % (name, index))
+            raise VariableError("%s '%s' has no item in index %d."
+                                % (type_name(variable, capitalize=True), name,
+                                   index))
 
-    def _parse_list_variable_index(self, index, support_slice=True):
+    def _parse_sequence_variable_index(self, index, support_slice=True):
         if ':' not in index:
             return int(index)
         if index.count(':') > 2 or not support_slice:
             raise ValueError
         return slice(*[int(i) if i else None for i in index.split(':')])
 
-    def _get_dict_variable_item(self, name, variable, key):
-        key = self.replace_scalar(key)
+    def _get_subscriptable_variable_item(self, name, variable, item):
+        """Gets item from a subscriptable variable.
+
+        This variable can be a dictionary for example, but also a custom class
+        that implements __getitem__(). If the item is not a variable, the
+        subscriptable variable gets treated as a sequence, which means
+        that the item is allowed to be an index/slice given as a string.
+        """
+        item = self.replace_scalar(item)
         try:
-            return variable[key]
+            return variable[item]
         except KeyError:
-            raise VariableError("Dictionary '%s' has no key '%s'."
-                                % (name, key))
-        except TypeError as err:
-            raise VariableError("Dictionary '%s' used with invalid key: %s"
-                                % (name, err))
+            raise VariableError("%s '%s' has no item '%s'."
+                                % (type_name(variable, capitalize=True),
+                                   name, item))
+        except Exception as err:
+            # Try to treat custom class as a Sequence but don't raise error
+            # on failure
+            try:
+                return self._get_sequence_variable_item(name, variable, item)
+            except:
+                pass
```

Comment:
I don't think it is a good idea to first try dictionary access and then sequence access if that fails. It's just too complicated and error prone compared to the benefits. Better to decide which access to use based on the variable type. Just need to decide what algorithm to use for that. Let's discuss more about that in the comments of the original issue or on Slack.